### ISSUE TITLE
Allow custom IServiceProvider on OWIN

### ISSUE THREAD

**Issue by tomasherceg** [riganti/dotvvm#628]
Currently, DotVVM creates `IServiceProvider` on its own in the OWIN version and it is not possible to replace it by a custom implementation. It makes difficult to use DotVVM with a custom IoC/DI container - you need to supply your own `IViewModelLoader` and now also `IStaticCommandServiceLoader`. 
Also, you need to take care of scopes on your own. 

We should add an option to return your own `IServiceProvider` like you can on ASP.NET Core, which would allow nicer integration with IoC/DI containers. 
We would also be able to deprecate `IViewModelLoader` and `IStaticCommandServiceLoader`.

**Comment by exyi** [comment]
There is no reason to deprecate `IViewModelLoader`, it can be used for custom initialization of the ViewModel if you don't like the default constructor injection mechanism.

**Comment by tomasherceg** [comment]
OK, but we should at least add `context` parameter to `DisposeViewModel` method - it's not there and it could be useful in some scenarios.

**Comment by exyi** [comment]
> we should at least add context parameter to DisposeViewModel method - it's not there and it could be useful in some scenarios.

That would be a breaking change, and I think it's totally not worth it. You can always mark the service as per-request and get the context from `IServiceProvider`.

**Comment by tomasherceg** [comment]
How would you get the `IServiceProvider` instance (for the current request) in `DisposeViewModel`? 
Yes, you can use `HttpContext.Current.GetOwinContext().GetDotvvmContext().Services`, but that's not nice. In ASP.NET Core, there is `IHttpContextAccessor`, but I am not aware of anything like that in OWIN.

**Comment by exyi** [comment]
> How would you get the `IServiceProvider` instance (for the current request) in `DisposeViewModel`? 

Constructor injection.

**Comment by exyi** [comment]
> ViewModelLoader is registered as a singleton, so you would need to reconfigure it to scoped or transient.

Yes, you have to register your own instance anyway, this is not a problem.

**Comment by tomasherceg** [comment]
No, this is just stupid.

**Comment by exyi** [comment]
>No, this is just stupid.
What? Using a DI container? Having it singleton by default? Your proposals?

**Comment by tomasherceg** [comment]
Changing lifetime of instances to something non-standard. 
Maybe in this case, it's not such a big deal, but I would be careful with that in general.
